### PR TITLE
[bazel] Add "dual" library targets for easier device/host code sharing

### PR DESCRIPTION
**Pull request by mcy**
This is currently a draft that illustrates how such a macro can help simplify a lot of our shared code. We can now define a single target for a "mockable" library, rather than three:

```c
opentitan_dual_cc_library(
    name = "boot_data",
    srcs = {"device": ["boot_data.c"]},
    hdrs = {
        "shared": ["boot_data.h"],
        "host": ["mock_boot_data.h"],
    },
    deps = {
        "shared": [
            ":error",
            "//sw/device/lib/base",
            "//sw/device/silicon_creator/lib/drivers:hmac",
            "//sw/device/silicon_creator/lib/drivers:lifecycle",
        ],
        "device": [
            "//hw/top_earlgrey/sw/autogen:top_earlgrey",
            "//sw/device/silicon_creator/lib/drivers:flash_ctrl",
        ],
        "host": [
            "//sw/device/lib/base/testing:global_mock",
            "//sw/device/silicon_creator/testing:mask_rom_test",
            "@googletest//:gtest",
        ],
    },
)
```

Unfortunately, `testonly` is not configurable, so `:boot_data` is not `testonly` anymore. However, Bazel's docs suggest this is WAI:
> Because testonly is enforced at build time, not run time, and propagates virally through the dependency tree, it should be applied judiciously. For example, stubs and fakes that are useful for unit tests may also be useful for integration tests involving the same binaries that will be released to production, and therefore should probably not be marked testonly. Conversely, rules that are dangerous to even link in, perhaps because they unconditionally override normal behavior, should definitely be marked testonly.

I want to get some feedback on this design before I do a full-on conversion of all of mask-rom's targets.

### CODE REVIEW

**Comment by mcy** [comment]
...GitHub  failed to make this a draft PR and now I can't undo it? Oops.

**Comment by mcy** [comment]
Never mind GH's UI is bad.

File: rules/opentitan.bzl [hunk 1]
```diff
@@ -802,3 +802,46 @@ def opentitan_functest(
         name = name,
         tests = all_tests,
     )
+
+def _get_or_empty(dict, key):
```
Comment (drewmacrae) [inline rules/opentitan.bzl:806]: Can you check the type of dict so you don't get passed a list of deps and ignore it all quietly here?

**Review by drewmacrae — COMMENTED** [review]
I like it a lot. In addition to trying to throw a useful warning if we're ever assuming a list is a dictionary in _get_or_empty, 

- It's probably worth enabling lists of srcs deps or headers that are by default shared so this doesn't prevent a more concise expression of simple rules. 
- I'd be happy to migrate the other sources, especially once we have some test coverage for them in CI if we like the idea of moving in this direction to automate and simplify things.

**Comment by drewmacrae** [comment]
It looks like the target you list with a dual_library cc rule is handled appropriately in the draft PR @cfrantz is working on here https://github.com/lowRISC/opentitan/pull/11416 Which isn't so friendly with Meson but involves only the built-in rules.

I've personally been focused on build system reimplementation more than what's going on in the source, so I have more trouble understanding what Chris is doing in that draft, but this PR may be working on a solved problem.

**Comment by mcy** [comment]
https://github.com/lowRISC/opentitan/pull/11776 is a complete prototype.

**Pull request closed without merging**